src/array.rs

```diff
@@ -1,98 +1,21 @@
-/// one atomic value (i.e. a single Miniconf item).
-///
-/// The type you should use depends on what data is contained in your array. If your array contains
-/// `Miniconf` items, you can (and often want to) use [`Array`] and the `#[miniconf(defer)]` attribute.
+/// The type to use depends on what data is contained in your array. If your array contains
+/// `Miniconf` items, you can (and often want to) use `D >= 2`.
 /// However, if each element in your list is individually configurable as a single value (e.g. a list
-/// of `u32`), then you must use a standard [`[T; N]`](array) array but you may optionally
-/// `#[miniconf(defer)]` access to individual indices.
-///
-/// # Construction
-///
-/// An `Array` can be constructed using [`From<[T; N]>`](From)/[`Into<miniconf::Array>`]
-/// and the contained value can be accessed through [`Deref`]/[`DerefMut`].
-#[derive(Clone, Copy, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
-#[repr(transparent)]
-pub struct Array<T, const N: usize>([T; N]);
-
-impl<T, const N: usize> Deref for Array<T, N> {
-    type Target = [T; N];
-
-    fn deref(&self) -> &Self::Target {
-        &self.0
-    }
-}
-
-impl<T, const N: usize> DerefMut for Array<T, N> {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.0
-    }
-}
-
-impl<T: Default + Copy, const N: usize> Default for Array<T, N> {
-    fn default() -> Self {
-        Self([T::default(); N])
-    }
-}
-
-impl<T, const N: usize> From<[T; N]> for Array<T, N> {
-    fn from(x: [T; N]) -> Self {
-        Self(x)
-    }
-}
-
-impl<T, const N: usize> AsRef<[T; N]> for Array<T, N> {
-    fn as_ref(&self) -> &[T; N] {
-        self
-    }
-}
-
-impl<T, const N: usize> AsMut<[T; N]> for Array<T, N> {
-    fn as_mut(&mut self) -> &mut [T; N] {
-        self
-    }
-}
-
-impl<T, const N: usize> IntoIterator for Array<T, N> {
-    type Item = T;
-    type IntoIter = <[T; N] as IntoIterator>::IntoIter;
-    fn into_iter(self) -> Self::IntoIter {
-        self.0.into_iter()
-    }
-}
-
-impl<'a, T, const N: usize> IntoIterator for &'a Array<T, N> {
-    type Item = <&'a [T; N] as IntoIterator>::Item;
-    type IntoIter = <&'a [T; N] as IntoIterator>::IntoIter;
-    fn into_iter(self) -> Self::IntoIter {
-        self.iter()
-    }
-}
-
-impl<'a, T, const N: usize> IntoIterator for &'a mut Array<T, N> {
-    type Item = <&'a mut [T; N] as IntoIterator>::Item;
-    type IntoIter = <&'a mut [T; N] as IntoIterator>::IntoIter;
-    fn into_iter(self) -> Self::IntoIter {
-        self.iter_mut()
-    }
-}
-
-impl<T, const N: usize> From<Array<T, N>> for [T; N] {
-    fn from(x: Array<T, N>) -> Self {
-        x.0
-    }
-}
+/// of `u32`), then you must use `D = 1` or `D = 0` if all items are to be accessed simultaneously.
+/// For e.g. `[[T; 2]; 3] where T: Miniconf<3>` you may want to use `D = 5` (note that `D <= 2`
+/// will also work if `T: Serialize + DeserializeOwned`).
```

Comment:
This feels unfortunate to me because it means the user of miniconf then needs to somehow know the internal depth of some struct where they called `#[derive(Miniconf)]` and the depth was auto-calculated for them.

If I were writing i.e. Booster's settings tree, I'm not sure what depth I should actually use for the various channel settings arrays without having to do a relatively complex analysis. Is there any way we could save the user from this and make it "Just work"?

Reply:
I don't see a way to make it automatic.
The error message however is very convenient and tells you exactly the value it wants.
So during development I don't expect this to be inconvenient.
The fact that all the fragility due to the iterator state size and the talk about its constraints etc is gone is a major plus though.

Reply:
Ah, so if you do the wrong value, the compiler should guide you towards the right one? That does sound incredibly convenient. Is that also the case for the `MqttClient` type parameter?

Reply:
Iirc yes.